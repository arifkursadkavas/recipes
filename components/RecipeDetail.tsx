import * as React from 'react';
import Tag from './Tag';

const RecipeDetail: React.FC<RecipeDetailProps> = (props) => {
  const { title, imageUrl, chef, description, tags } = props;

  const formattedDesc = replaceRegex(description);

  return (
    <div className="recipeContainer">
      <img
        className="recipeImage"
        src={imageUrl}
        alt={title}
        width="787px"
        height="524px"
      ></img>
      <div className="recipeInfo">
        <div className="recipeTitle">{title}</div>
        <div
          className="recipeDescription"
          dangerouslySetInnerHTML={{ __html: formattedDesc }}
        ></div>
        <div className="recipeChef">Chef {chef ? chef : 'Anonymous'}</div>
        <div className="recipeTags">
          <div className="tagList">
            {tags.map((tag) => (
              <Tag text={tag.name} key={tag.name}></Tag>
            ))}
          </div>
        </div>
      </div>
    </div>
  );
};

//TODO:: Need to know the general patterns here. Below applies to only current data.
const replaceRegex = (str: string): string => {
  const starRegex = new RegExp(/\*(.*?)\*/, 'ig');
  str = str.replace(starRegex, '<span class="starClass">$1</span>');

  const usRegex = new RegExp(/_{2}(.*?)_{2}/, 'ig');

  str = str.replace(usRegex, '<span class="ucClass">$1</span>');

  const videoRegex = new RegExp(/\[VIDEO]\((.*?)\)/, 'ig');

  str = str.replace(videoRegex, '<a class="linkClass" href="$1">$1</a>');

  return str;
};

interface RecipeDetailProps {
  title: string;
  imageUrl: string;
  chef: string;
  description: string;
  tags: Array<{ name: string }>;
}

export default RecipeDetail;
