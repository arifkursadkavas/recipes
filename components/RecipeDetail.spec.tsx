import * as React from 'react';
import { mount, shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import RecipeDetail from './RecipeDetail';

describe('Components', () => {
  describe('RecipeDetail', () => {
    test('snapshot renders', () => {
      const component = renderer.create(
        <RecipeDetail
          title="Title"
          imageUrl="ImageUrl"
          chef="ChefName"
          description="Description"
          tags={[{ name: 'tag1' }, { name: 'tag2' }]}
        />
      );
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});

describe('Components', () => {
  describe('RecipeDetail', () => {
    it('should render title without throwing an error', function () {
      const wrap = mount(
        <RecipeDetail
          title="Title"
          imageUrl="ImageUrl"
          chef="ChefName"
          description="Description"
          tags={[{ name: 'tag1' }, { name: 'tag2' }]}
        />
      );
      expect(wrap.find('.recipeTitle').text()).toBe('Title');
    });
  });
});

describe('Components', () => {
  describe('RecipeDetail', () => {
    it('should render description without throwing an error', function () {
      const wrap = mount(
        <RecipeDetail
          title="Title"
          imageUrl="ImageUrl"
          chef="ChefName"
          description="Description"
          tags={[{ name: 'tag1' }, { name: 'tag2' }]}
        />
      );
      expect(wrap.find('.recipeDescription').text()).toBe('Description');
    });
  });
});

describe('Components', () => {
  describe('RecipeDetail', () => {
    it('should render imageUrl without throwing an error', function () {
      const wrap = mount(
        <RecipeDetail
          title="Title"
          imageUrl="ImageUrl"
          chef="ChefName"
          description="Description"
          tags={[{ name: 'tag1' }, { name: 'tag2' }]}
        />
      );
      expect(wrap.find('.recipeImage').prop('src')).toBe('ImageUrl');
    });
  });
});

describe('Components', () => {
  describe('RecipeDetail', () => {
    it('should render chef name without throwing an error', function () {
      const wrap = mount(
        <RecipeDetail
          title="Title"
          imageUrl="ImageUrl"
          chef="ChefName"
          description="Description"
          tags={[{ name: 'tag1' }, { name: 'tag2' }]}
        />
      );
      expect(wrap.find('.recipeChef').text()).toBe('Chef ChefName');
    });
  });
});

describe('Components', () => {
  describe('RecipeDetail', () => {
    it("should render chef name as 'Anonymous' when it is not supplied", function () {
      const wrap = mount(
        <RecipeDetail
          title="Title"
          imageUrl="ImageUrl"
          chef=""
          description="Description"
          tags={[{ name: 'tag1' }, { name: 'tag2' }]}
        />
      );
      expect(wrap.find('.recipeChef').text()).toBe('Chef Anonymous');
    });
  });
});

describe('Components', () => {
  describe('RecipeDetail', () => {
    it('should render Tag components without throwing an error', function () {
      const wrap = shallow(
        <RecipeDetail
          title="Title"
          imageUrl="ImageUrl"
          chef="ChefName"
          description="Description"
          tags={[{ name: 'tag1' }, { name: 'tag2' }]}
        />
      );
      expect(wrap.find('Tag')).toHaveLength(2);
    });
  });
});

describe('Components', () => {
  describe('RecipeDetail', () => {
    it("should render Tag component with 'text' property without throwing an error", function () {
      const wrap = shallow(
        <RecipeDetail
          title="Title"
          imageUrl="ImageUrl"
          chef="ChefName"
          description="Description"
          tags={[{ name: 'tag1' }]}
        />
      );
      expect(wrap.find('Tag').first()).toHaveProperty('text');
    });
  });
});

describe('Components', () => {
  describe('RecipeDetail', () => {
    it("should render Tag component with 'text' property value without throwing an error", function () {
      const wrap = shallow(
        <RecipeDetail
          title="Title"
          imageUrl="ImageUrl"
          chef="ChefName"
          description="Description"
          tags={[{ name: 'tag1' }]}
        />
      );
      expect(wrap.find('Tag').first().prop('text')).toEqual('tag1');
    });
  });
});
