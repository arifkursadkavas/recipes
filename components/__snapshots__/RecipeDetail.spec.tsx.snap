// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components RecipeDetail snapshot renders 1`] = `
<div
  className="recipeContainer"
>
  <img
    alt="Title"
    className="recipeImage"
    height="524px"
    src="ImageUrl"
    width="787px"
  />
  <div
    className="recipeInfo"
  >
    <div
      className="recipeTitle"
    >
      Title
    </div>
    <div
      className="recipeDescription"
    >
      Description
    </div>
    <div
      className="recipeChef"
    >
      Chef 
      ChefName
    </div>
    <div
      className="recipeTags"
    >
      <div
        className="tagList"
      >
        <div
          className="tagContainer"
        >
          <span
            className="tagText"
          >
            tag1
          </span>
        </div>
        <div
          className="tagContainer"
        >
          <span
            className="tagText"
          >
            tag2
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
